### PR TITLE
Retry unsuccessful bulk uploads.

### PR DESCRIPTION
**Pull request by aldenstpage**
## Description
During bulk indexing, Elasticsearch can become overwhelmed with writes and reject indexing requests. There could also be rare cases where the upstream data is malformed in a way that prevents the indexer from uploading a particular chunk of data. Either case causes the indexer worker to stop indexing its batch.

If a bulk indexing request is rejected for any reason, we will retry with exponential backoff. If `max_attempts` is exceeded, this batch will be skipped and we will continue indexing the rest of the data.

Once we have some more detailed logs, we will be able to implement better handling of malformed rows and discard them from the upload batch.

### CODE REVIEW

File: ingestion_server/ingestion_server/indexer.py [hunk 1]
```diff
@@ -213,6 +214,30 @@ def _index_table(self, table, dest_idx=None):
...
+        # Initial time to wait between indexing attempts
+        # Grows exponentially
+        cooloff = 5
+        while True:
+            try:
+                deque(helpers.parallel_bulk(self.es, es_batch, chunk_size=400))
+            except elasticsearch.ElasticsearchException:
+                # Something went wrong during indexing.
+                log.warning(
+                    "Elasticsearch rejected bulk query. We will retry in a"
```
Comment (brenoferreira) [inline ingestion_server/ingestion_server/indexer.py:229]: Is it worth logging the attempt # and how much time will wait until the next attempt?
Are any details about the bulk query logged here, so you can know what query failed (not sure if that would go in `exc_info`)

File: ingestion_server/ingestion_server/indexer.py [hunk 2]
```diff
@@ -213,6 +214,30 @@ def _index_table(self, table, dest_idx=None):
...
+    def _bulk_upload(self, es_batch):
+        max_attempts = 4
+        attempts = 0
+        # Initial time to wait between indexing attempts
+        # Grows exponentially
+        cooloff = 5
+        while True:
+            try:
+                deque(helpers.parallel_bulk(self.es, es_batch, chunk_size=400))
+            except elasticsearch.ElasticsearchException:
```
Comment (brenoferreira) [inline ingestion_server/ingestion_server/indexer.py:226]: Is `ElasticsearchException` the only exception that could be thrown in this case?

File: ingestion_server/ingestion_server/indexer.py [hunk 3]
```diff
@@ -254,7 +279,10 @@ def replicate(self, table, dest_index, query):
                 num_docs = len(es_batch)
                 log.info('Pushing {} docs to Elasticsearch.'.format(num_docs))
                 # Bulk upload to Elasticsearch in parallel.
-                list(helpers.parallel_bulk(self.es, es_batch, chunk_size=400))
+                try:
+                    self._bulk_upload(es_batch)
+                except ValueError:
+                    log.error('Failed to index chunk.')
```
Comment (brenoferreira) [inline ingestion_server/ingestion_server/indexer.py:285]: Also not sure if the batch info goes on the log here for later reference

File: ingestion_server/ingestion_server/indexer.py (hunk 2 above)
Comment (aldenstpage) [reply ingestion_server/ingestion_server/indexer.py:226]: A more specific exception would be better, but there is no way to predict what is being thrown until we run on a real dataset. We should be able to better narrow it down at a future date based on the information in `exc_info`.

File: ingestion_server/ingestion_server/indexer.py (hunk 3 above)
Comment (aldenstpage) [reply ingestion_server/ingestion_server/indexer.py:285]: We'll have detailed information from exc_info

File: ingestion_server/ingestion_server/indexer.py (hunk 1 above)
Comment (aldenstpage) [reply ingestion_server/ingestion_server/indexer.py]: I think you're right that that could be useful. I added attempt number and the cooldown period to the logs.

exc_info will contain detailed information about why the bulk request failed.

File: ingestion_server/ingestion_server/indexer.py (hunk 2 above)
Comment (brenoferreira) [reply ingestion_server/ingestion_server/indexer.py:226]: Right. I thought that maybe there could be a more general exception handler, but if not, then it's fine

**Pull request merged**